src/NHibernate/Multi/FutureExtensions.cs

```diff
@@ -0,0 +1,272 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NHibernate
+{
+	public static partial class FutureExtensions
```

Comment:
In fact, these are `IQueryBatch` extensions. If `IQueryBatch` should be usable as a replacement for multi-queries and multi-criteria, these extensions only meant for futures will be quite polluting. Do we need them to be public and/or on `IQueryBatch`?

For me, the "futures yielding" extensions here only have internal usages.

I think public extensions here should instead yield back the `batch` for allowing chaining in user code while hiding the `IQueryBatchItem` encapsulation required for adding a query to the batch.

The "futures yielding" extensions should be made internal or moved to a `FutureHelper` class not extending `IQueryBatch` in order to avoid bloating its interface.